nd4j/src/main/scala/nexus/impl/nd4j/ND4JBackend.scala

```diff
@@ -0,0 +1,168 @@
+package nexus.impl.nd4j
+
+import nexus.algebra._
+import nexus.algebra.typelevel.util._
+import org.nd4j.linalg.factory.Nd4j
+import org.nd4j.linalg.ops.transforms.Transforms
+
+import scala.util._
+
+
+/**
+  * @author Andrey Romanov
+  */
+object ND4JBackendTensor extends IsRealTensorK[FloatTensor, Float]{
+  type H = UntypedFloatTensor
+  val H = UntypedND4JBackendTensor
+
+  val R = H.R
+
+  def newTensor[A](shape: Array[Int]) =
+    FloatTensor.fromFlatArray[A](Array.ofDim[Float](ShapeUtils.product(shape)), shape)
+
+  def newGaussianTensor[A](μ: Double, σ2: Double, shape: Array[Int]) = {
+    val r = new Random()
+    val σ = math.sqrt(σ2)
+    FloatTensor.fromFlatArray(Array.fill(ShapeUtils.product(shape))(((r.nextGaussian() - μ) * σ).toFloat), shape)
+  }
+
+  def get(x: FloatTensor[_], is: Array[Int]): Float = {
+    x.handle.getFloat(x.index(is))
+  }
+  def set(x: FloatTensor[_], is: Array[Int], v: Float) = x.update(is: _*)(v)
+  def zeroBy[A](x: FloatTensor[A]) = newTensor(x.shape)
+
+  def untype(x: FloatTensor[_]) = x
+  def typeWith[A](x: UntypedFloatTensor) = x.typeWith[A]
+
+  def fromDouble(d: Double) = d.toFloat
+  def fromFloat(f: Float) = f
+}
+
+
+object UntypedND4JBackendTensor extends IsUntypedRealTensor[UntypedFloatTensor, Float] {
+
+  type R = Float
+  val R = nexus.algebra.instances.Float32
+
+  def fromFlatArray(array: Array[R], shape: Array[Int]) =
+    new UntypedFloatTensor.Tensor(Nd4j.create(array, shape))
+
+  def mutable = true
+
+  def rank(x: UntypedFloatTensor) = x.rank
+  def shape(x: UntypedFloatTensor) = x.shape
+  def slice(x: UntypedFloatTensor, dim: Int, i: Int) = x.sliceUntyped(dim, i)
+
+
+  /*
+  we could use nd4s collection-like ops but it's broken
+  https://github.com/deeplearning4j/nd4s/issues/91
+   */
+  def map(x: UntypedFloatTensor)(f: Float => Float): UntypedFloatTensor = {
+    ???
+  }
+
+  def map2(x1: UntypedFloatTensor, x2: UntypedFloatTensor)(f: (Float, Float) => Float): UntypedFloatTensor = {
+    ???
+  }
+
+  def map3(x1: UntypedFloatTensor, x2: UntypedFloatTensor, x3: UntypedFloatTensor)(f: (Float, Float, Float) => Float): UntypedFloatTensor = {
+    ???
+  }
+
+  def inplace2(x1: UntypedFloatTensor, x2: UntypedFloatTensor)(f: (Float, Float) => Float): Unit = {
+    ???
+  }
+
+
+  def zeroBy(x: UntypedFloatTensor) = new UntypedFloatTensor.Tensor(Nd4j.zeros(x.shape: _*))
+
+
+  // Here and after we rely on build in Nd4j Ops instead of inefficient `map/map2/inplace2` implementation
```

Comment:
Please don't do `reshape` as this supports broadcasting on tensors. Nexus is completely typesafe -- No broadcasting should be allowd. Every broadcasting operation should be done explicitly using operators. For `add` it could just be `x1.add(x2)`. (directly using `INDArray` without using the `UntypedFloatTensor` wrapper).

Reply:
Okay, got it. Actually there are some difficulties with converting 1d arrays to INDArray since nd4j doesn't have explicit type for 1d array  `(d,)`, in nd4j it's either `(d,1)` or `(1,d)`. And this requires some extra thinking on how to align 1d arrays without using `reshape`s.

Reply:
Interesting. `Nd4j` doesn't support vectors (`Tensor[A]` in Nexus)?
Does it support scalar (0-dim) tensors (typed in Nexus as `Tensor[Unit]`)?

Reply:
seems like `nd4j` doesn't support scalar and vector tensors. So probably we need to wrap ND4J into custom class to handle it's type: scalar, vector or multidimentional.